### PR TITLE
Nullable enabled: LangaugePrimitives, PSObject,MshMemberInfo

### PR DESCRIPTION
**Pull request by powercode**
# PR Summary

`#nullable enable` in LanguagePrimitives.cs, MshObject.cs and MshMemberInfo.cs

Annotating with `?` and `!` to make the contract clear, and fixed a few issues, where `object.ToString` and `Type.Fullname`  are not guaranteed to return non-null strings.

## PR Context

Taking advantage of the C# 8 nullable reference types, for better clarity and less risk of null errors.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 1]
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) Microsoft Corporation.
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs:1]: @powercode, your last commit had 8 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Basic SysLog tests on Linux.Verifies logging level filtering works
```powershell
Expected $null, but got PSLogItem.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 256
256:         $items | Should -Be $null
```
SemanticVersion api tests.Error handling.Missing parts: 'NullString': ''
```powershell
Expected an exception, with FullyQualifiedErrorId 'PSArgumentNullException' to be thrown, but the FullyQualifiedErrorId was 'FormatException'. from /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:206 char:15
    +             { [SemanticVersion]::new($version) } | Should -Throw -Err ?
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 206
206:             { [SemanticVersion]::new($version) } | Should -Throw -ErrorId $errorId
```
Alias tests.Export-Alias literal path.for CSV
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for PS1
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for Null
```powershell
Expected strings to be the same, but they were different.
Expected length: 95
Actual length:   102
Strings differ at index 32.
Expected: '...ErrorNullN...'
But was:  '...ErrorEmpty...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs:1]: @powercode, your last commit had 7 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

SemanticVersion api tests.Error handling.Missing parts: 'NullString': ''
```powershell
Expected an exception, with FullyQualifiedErrorId 'PSArgumentNullException' to be thrown, but the FullyQualifiedErrorId was 'FormatException'. from /Users/runner/runners/2.168.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1:206 char:15
    +             { [SemanticVersion]::new($version) } | Should -Throw -Err ?
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 206
206:             { [SemanticVersion]::new($version) } | Should -Throw -ErrorId $errorId
```
Alias tests.Export-Alias literal path.for CSV
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for PS1
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for Null
```powershell
Expected strings to be the same, but they were different.
Expected length: 95
Actual length:   102
Strings differ at index 32.
Expected: '...ErrorNullN...'
But was:  '...ErrorEmpty...'
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Import-Alias literal path.for Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand'. from /Users/runner/runners/2.168.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1:110 char:45
    +                 { Import-Alias -LiteralPath $test.testFile -ErrorActi ?
    +                                             ~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/alias.tests.ps1: line 110
110:                 { Import-Alias -LiteralPath $test.testFile -ErrorAction SilentlyContinue } | Should -Throw -ErrorId $test.expectedError
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs:1]: @powercode, your last commit had 8 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

SemanticVersion api tests.Error handling.Missing parts: 'NullString': ''
```powershell
Expected an exception, with FullyQualifiedErrorId 'PSArgumentNullException' to be thrown, but the FullyQualifiedErrorId was 'FormatException'. from D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1:206 char:15
    +             { [SemanticVersion]::new($version) } | Should -Throw -Err ?
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 206
206:             { [SemanticVersion]::new($version) } | Should -Throw -ErrorId $errorId
```
Alias tests.Export-Alias literal path.for CSV
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for PS1
```powershell
Expected <empty>, but got $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Export-Alias literal path.for Null
```powershell
Expected strings to be the same, but they were different.
Expected length: 95
Actual length:   102
Strings differ at index 32.
Expected: '...ErrorNullN...'
But was:  '...ErrorEmpty...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 57
57:                     $exportAliasError.FullyqualifiedErrorId | Should -Be $test.expectedError
```
Alias tests.Import-Alias literal path.for Null
```powershell
Expected an exception, with FullyQualifiedErrorId 'ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.ImportAliasCommand'. from D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1:110 char:45
    +                 { Import-Alias -LiteralPath $test.testFile -ErrorActi ?
    +                                             ~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\alias.tests.ps1: line 110
110:                 { Import-Alias -LiteralPath $test.testFile -ErrorAction SilentlyContinue } | Should -Throw -ErrorId $test.expectedError
```

**Comment by iSazonov** [comment]
@powercode I really like it and I fully support it. But there's a problem. I started this process in #10693 without progress because MSFT team has no internal resources. 
Thus, if we want to make a progress, we need a plan, approval of MSFT team and, most importantly, _several active participants_ for contributions and code reviews for some months.

If you are ready to participate, we can ask @vexx32 and some other developers who is interested in binary module development.

**Comment by powercode** [comment]
@iSazonov I have used this feature for a while both in my personal and professional projects, and it is now beyond doubt something that makes me write better code. 

I tend to work on some feature I want, and refactor the relevant code to a good state before I start, so this was just a prequel to work on the parameter binder.

I see the problem, but I think this is something that is so important for code quality that it may be wise to have it at a higher priority.

It is somewhat time consuming and hard to review these - you really have to understand the semantics of subsystem to be able to get it right.

But let's try to get @vexx32 on board :) 

How can I be of help here?

**Comment by powercode** [comment]
https://github.com/dotnet/coreclr/pull/23466 discussion around object.ToString being `string?`.

Our `NullString.ToString` is such a case.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 1 above)
Comment (powercode) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs:1]: Fixed by #f44e12d

Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs:1]: @powercode, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Int*ro' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'New-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Invoke-CimMethod -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimClass -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs:1]: @powercode, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Int*ro' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'New-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Invoke-CimMethod -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimClass -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```

**Comment by vexx32** [comment]
@iSazonov @powercode from my perspective, the code has larger style issues in many areas anyway. I would tend to prefer these kinds of issues be fixed first (and in some cases cleaning up the code would make nullable annotations simpler to decide on), but I do agree the value of having nullable reference types enabled is a little higher since it is picked up by analysis tools more easily.

I'm always in for some code cleanup. 😁

**Comment by SeeminglyScience** [comment]
@vexx32 nullability annotations is a lot more than style.  It significantly reduces accidental NRE exceptions in-project, and also when public API's are annotated it provides a way to tell if something can be null without having to look up documentation.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 1 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs:1]: @powercode, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimInstance -Namespace root/Int*ro' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'New-CimInstance -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Invoke-CimMethod -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```
TabCompletion.CIM cmdlet completion tests.CIM cmdlet input 'Get-CimClass -Namespace root/Inter' should successfully complete
```powershell
Expected strings to be the same, but they were different.
String lengths are both 12.
Strings differ at index 5.
Expected: 'root/Interop'
But was:  'root/interop'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 1252
1252:             $res.CompletionMatches[0].CompletionText | Should -BeExactly $expected
```

**Comment by adityapatwardhan** [comment]
@powercode please have a look at merge conflict.

**Comment by powercode** [comment]
rebased on master

File: src/System.Management.Automation/engine/MshObject.cs [hunk 2]
```diff
@@ -1552,19 +1540,19 @@ public string ToString(string format, IFormatProvider formatProvider)
...
-            string result;
+            string? result;
             try
             {
                 result = this.ToString();
             }
             catch (ExtendedTypeSystemException)
             {
-                result = this.BaseObject.GetType().FullName;
+                result = this.BaseObject.GetType().FullName ?? string.Empty;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshObject.cs:1555]: Please revert the unrelated change.

File: src/System.Management.Automation/engine/MshObject.cs [hunk 3]
```diff
@@ -1687,8 +1675,9 @@ public int CompareTo(object obj)
             }
             catch (ArgumentException e)
             {
+                var objStr = obj == null ? string.Empty : PSObject.AsPSObject(obj).ToString();
                 throw new ExtendedTypeSystemException("PSObjectCompareTo", e,
-                    ExtendedTypeSystem.NotTheSameTypeOrNotIcomparable, this.PrivateToString(), PSObject.AsPSObject(obj).ToString(), "IComparable");
+                    ExtendedTypeSystem.NotTheSameTypeOrNotIcomparable, this.PrivateToString(), objStr, "IComparable");
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshObject.cs:1680]: Why do we need this?

File: src/System.Management.Automation/engine/MshObject.cs [hunk 4]
```diff
@@ -1846,8 +1835,8 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
         /// PSObject to be used while replicating the owner for PSMemberSet
         /// </param>
         /// <returns></returns>
-        internal static object GetNoteSettingValue(PSMemberSet settings, string noteName,
-            object defaultValue, Type expectedType,
+        internal static object? GetNoteSettingValue(PSMemberSet? settings, string noteName,
+            object? defaultValue, Type expectedType,
             bool shouldReplicateInstance, PSObject ownerObject)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshObject.cs:1840]: Looking how the method is used we can make it non-nullable:
```suggestion
        internal static object GetNoteSettingValue(PSMemberSet? settings, string noteName,
            object defaultValue, Type expectedType,
            bool shouldReplicateInstance, PSObject ownerObject)
```

File: src/System.Management.Automation/engine/MshObject.cs [hunk 5]
```diff
@@ -1875,16 +1864,16 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
...
 
-            TypeTable typeTable = backupTypeTable ?? this.GetTypeTable();
+            TypeTable? typeTable = backupTypeTable ?? this.GetTypeTable();
             if (typeTable != null)
             {
-                PSMemberSet standardMemberSet = TypeTableGetMemberDelegate<PSMemberSet>(this,
+                PSMemberSet? standardMemberSet = TypeTableGetMemberDelegate<PSMemberSet>(this,
                     typeTable, TypeTable.PSStandardMembers);
-                result = (int)GetNoteSettingValue(standardMemberSet, TypeTable.SerializationDepth, 0, typeof(int), true, this);
+                result = (int)GetNoteSettingValue(standardMemberSet, TypeTable.SerializationDepth, 0, typeof(int), true, this)!;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshObject.cs:1876]: With my previous comment:
```suggestion
                result = (int)GetNoteSettingValue(standardMemberSet, TypeTable.SerializationDepth, 0, typeof(int), true, this);
```

File: src/System.Management.Automation/engine/MshObject.cs [hunk 6]
```diff
@@ -1912,23 +1901,23 @@ internal PSPropertyInfo GetStringSerializationSource(TypeTable backupTypeTable)
...
-            TypeTable typeTable = backupTypeTable ?? this.GetTypeTable();
+            TypeTable? typeTable = backupTypeTable ?? this.GetTypeTable();
             if (typeTable != null)
             {
-                PSMemberSet standardMemberSet = TypeTableGetMemberDelegate<PSMemberSet>(this,
+                PSMemberSet? standardMemberSet = TypeTableGetMemberDelegate<PSMemberSet>(this,
                     typeTable, TypeTable.PSStandardMembers);
                 result = (SerializationMethod)GetNoteSettingValue(standardMemberSet,
-                        TypeTable.SerializationMethodNode, TypeTable.DefaultSerializationMethod, typeof(SerializationMethod), true, this);
+                        TypeTable.SerializationMethodNode, TypeTable.DefaultSerializationMethod, typeof(SerializationMethod), true, this)!;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshObject.cs:1914]:
```suggestion
                        TypeTable.SerializationMethodNode, TypeTable.DefaultSerializationMethod, typeof(SerializationMethod), true, this);
```

File: src/System.Management.Automation/engine/MshObject.cs [hunk 7]
```diff
@@ -2132,13 +2121,14 @@ internal class PSDynamicMetaObject : DynamicMetaObject
             internal PSDynamicMetaObject(Expression expression, PSObject value)
                 : base(expression, BindingRestrictions.Empty, value)
             {
+                Diagnostics.Assert(value != null, "Should be verified by callers.");
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshObject.cs:2124]: Does this come from usages the ctor/class? If so please make runtime check and throw.

File: src/System.Management.Automation/engine/MshObject.cs [hunk 8]
```diff
@@ -2615,22 +2605,18 @@ public static string Type(PSObject instance)
...
-            XmlNode node = (XmlNode)instance?.BaseObject;
-            if (node == null)
-            {
-                return string.Empty;
-            }
-
-            return node.LocalName;
+            return instance?.BaseObject is XmlNode node
+                ? node.LocalName
+                : string.Empty;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/MshObject.cs:2610]: Please revert the unrelated change.

**Review by iSazonov — COMMENTED** [review]
I reviewed only MshObject.cs file. Please move two other files in follow PRs - this speed up the review.
_Also please remove all unrelated changes - they change behaviors and add a risk of regressions._
